### ISSUE TITLE
language specific options feature request

### ISSUE THREAD

**Issue by Reobos** [SublimeLinter/SublimeLinter-clang#3]
I added two new options in pull request #1 so that I can specify cflags separately for c++ and c sources (does not break backward compatibility). This is especially important for projects that have both c and c++ source files. Rather than having to change the linter settings every time a different source file is opened in a project, it's much easier to "set it and forget it". 

It's common to specify the c/c++ language standard flags and is a perfect example for this added functionality. For example, '-std=c99' for c source files and '-std=c++11' for c++ source files:
```
"SublimeLinter":
{
    "linters":
    {
        "clang": {
            "extra_cflags": "-std=c99",
            "extra_cxxflags": "-std=c++11",
            "extra_flags": "-Wall -I${project_path}/foo",
            "include_dirs": [
                "${project_path}/3rdparty/bar/include",
                "${project_path}/3rdparty/baz"
            ]
        }
    }
}
```

**Comment by kaste** [comment]
I wonder if we should just export two linters here. (You would still install one plugin of course.) Then we would have two linter settings.

**Comment by Reobos** [comment]
I'm not sure what you mean by two linters? Is that easy to do?

If you look at the [linter.py source](https://github.com/SublimeLinter/SublimeLinter-clang/blob/b9aa3f12356f647b1bb78cf81a5609eb134d2f35/linter.py#L59), this linter actually already sets different flags for C/C++ sources. It adds ' -x c ' for C files and ' -x c++ ' for C++ files. This makes it very easy to add the settings flags 'extra_cflags' and 'extra_cxxflags', which is done in pull request #1

**Comment by kaste** [comment]
When we define two classes named clang and cpluslang or whatever we would get different settings for them. Since both would define their own syntaxes we wouldn't need to check the syntax in the code. I just thought maybe that's the right thing to do. Otherwise and for the feature per se your PR is 👍

**Comment by jfcherng** [comment]
`SublimeLinter-gcc` implements this too, but in a more general way. It allows user too add prefixed `c_` or `c++_` to any settings to claim C/C++-specific settings. I could even use `SublimeLinter-gcc` to utilize `clang` and `clang++` since major flags are quite the same with `gcc`.

```javascript
{
    "linters":
    {
        "gcc": {
            "disable": false,
            // C-specific settings
            "c_executable": "clang",
            "c_extra_flags": [
                "-fsyntax-only",
                "-std=c99",
            ],
            // C++-specific settings
            "c++_executable": "clang++",
            "c++_extra_flags": [
                "-fsyntax-only",
                "-std=c++11",
            ],
            // include_dirs for both C and C++
            "include_dirs": [
                "${project_folder}/include",
                "/usr/local/include",
            ],
        },
    },
}
```

`SublimeLinter-clang` could takes advantages/codes from `SublimeLinter-gcc` (or even be a fork, imo). 
For example, [string template](https://github.com/SublimeLinter/SublimeLinter-gcc/blob/master/linter.py#L102) is lots more clear then [string concat](https://github.com/SublimeLinter/SublimeLinter-clang/blob/master/linter.py#L59-L70).

---

References:
- https://github.com/SublimeLinter/SublimeLinter-gcc#settings
- https://github.com/SublimeLinter/SublimeLinter-gcc/blob/551e00b67fac91c36308675d5742cc94a3204233/linter.py#L147-L160

**Comment by kaste** [comment]
Hm, I think if you have 

```
class Clang(Linter):
    defaults = {
        'selector': 'source.c'
    }

    regex = (
        r'<stdin>:(?P<line>\d+):'
        r'((?P<col>\d*): )?'# column number, colon and space are only applicable for single line messages
        # several lines of anything followed by
        # either error/warning/note or newline (= irrelevant backtrace content)
        # (lazy quantifiers so we don’t skip what we seek)
        r'(.*?((?P<error>error)|(?P<warning>warning|note)|\r?\n))+?'
        r': (?P<message>.+)'# match the remaining content of the current line for output
    )
    multiline = True
    
    cmd = (
        'clang -fsyntax-only ',
        '-fno-caret-diagnostics -Wall ',
        '-x c',
        '${args}',
        '-'
    )

```

That is probably all you need. Then you do the same for cplus, and you would have two linter settings sections.

**Comment by kaste** [comment]
Explainer: `${args}` is a special setting available for all linters. Users can set this per linter:

```
 	linters: {
		clang: {
			args: [
				'-I foo/bar',
				'-I foo/baz',
			]
		}
	}
```

If you can guess good includes, it would be of course necessary to implement `cmd` as a function.

**Comment by Reobos** [comment]
The sublimelinter-gcc approach takes care in providing features, settings, and compatibility with sublimelinter 3. One thing I think would be useful is adding the -fsyntax-only flag so you don't have to output  a temp (garbage) file. Having the cmd being modified in settings is unusual, but does make sense for very similar compiler outputs. What is a good way to have multiple executables for similar compilers? Perhaps a linter for each of [clang, clang++, gcc, g++, avr-gcc, avr-gcc++] like @kaste suggested or a setting like sublimelinter-gcc?

@kaste I like your simplified approach that actually applies the bare minimum on top of what sublimelinter already does, without sacrificing customizability. However, I do like having contextual settings specific to the compiler to keep things more clear (kind of like an IDE)

**Comment by kaste** [comment]
If it's really just the executable, you can (in the next SublimeLinter version 4.1) set `linters > clang > executable: 'my/other/executable'`. The most important and unique thing to implement for a linter class is the error/pattern matching really.

But @jfcherng is the 🤴 here for the c languages. I just wanted to show that it is probably not necessary to abstract over the small differences for the aforementioned compilers and IMO it is a good solution to register different linters with SublimeLinter bc then you get different settings objects too.

**Comment by Reobos** [comment]
I just installed  SublimeLinter-gcc. I like it! Thanks @jfcherng

**Comment by Reobos** [comment]
Fixed with Cleanup #7
